TripalCultivate/TripalCultivate-Phenotypes
# G4.20 genus project service

**Issue https://github.com/TripalCultivate/TripalCultivate-Phenotypes/issues/10** - [Genus Project Service](https://github.com/TripalCultivate/TripalCultivate-Phenotypes/issues/20)

## Motivation
This service is the upgraded version of [Analyzed Phenotypes Genus API](https://github.com/UofS-Pulse-Binfo/analyzedphenotypes/blob/7.x-3.x/api/analyzedphenotypes.genus.api.inc) with a focus on the new D10 functionality.



## What does this PR do?
Service will handle setting of genus to project and resolving genus of a project. Mapping of genus is done using projectprop table where type_id is the configuration value of the term genus.

## Testing

### Automated Testing
*Please describe each automated test this PR creates and provide a list of the assertions it makes using casual language.*
*Do not just say things like "asserts the array is not empty" but rather say "Ensures that the return value of method X with these parameters is not an empty array".*

ClassName->ServiceGenusProjectTest: 
- Create test records in chado.organism and chado.project.
- Create relationship record in chado.projectprop between genus and project.
- Create genus configuration entry (genus: trait, unit, method, db and crop ontology).
- Tests that service can retrieve active genus - genus that have been configured.
- Test to retrieve the genus of a project.
- Test to replace/update a project with genus.

### Manual Testing
*Describe in detail how someone should manually test this functionality.*
*Make sure to include whether they need to build a docker from scratch, create any records, etc.*

1. Install and enable Tripal Cultivate Phenotypes module and switch branch to g4.01-Genus-Project-Service
Note: make sure to have a test organism and project records and an entry in projectprop table (of type the term in configuration corresponding to genus) relating a genus to a project, in your instance.
2. Configure terms and ontology of the module and create a genus-ontology configuration entry.
3. Using php in your host/devel/php
4. test: 
  $service = \Drupal::service('trpcultivate_phenotypes.genus_project');
  $service->getActiveGenus(); // Should list all genus that have been configured.
  $service->getGenusOfProject(project id number); // Should return a genus of a project.
  $service->setGenusToProject(project id number, genus); // Should set a genus to a project.
